rcl/src/rcl/domain_id.c

```diff
@@ -40,10 +41,14 @@ rcl_get_default_domain_id(size_t * domain_id)
       get_env_error_str);
     return RCL_RET_ERROR;
   }
-  if (ros_domain_id) {
+  if (ros_domain_id && strcmp(ros_domain_id, "") != 0) {
     unsigned long number = strtoul(ros_domain_id, NULL, 0);  // NOLINT(runtime/int)
-    if (number == ULONG_MAX) {
-      RCL_SET_ERROR_MSG("failed to interpret ROS_DOMAIN_ID as integral number");
+    if (number == 0UL && strspn(ros_domain_id, "0") != strlen(ros_domain_id)) {
```

Comment:
I didn't remember that `strtoul` can return `0` on error. That's pretty weird ...

Reply:
[`strtoul()`](https://en.cppreference.com/w/cpp/string/byte/strtoul) returns `ULONG_MAX` in case of an overflow (depending on the size of `unsigned long` on the specific platform). But an invalid character is not an overflow and `strtoul()` just stops parsing. You would need to add a `char ** str_end` argument instead of `NULL` as the second argument and check whether `str_end` points to the terminating null character.

See also my comment in https://github.com/ros2/rclcpp/pull/1188#discussion_r443083058. `rclcpp` obviously has its own implementation of `rcl_get_default_domain_id()` and does not use this one.

Reply:
> You would need to add a char ** str_end argument instead of NULL as the second argument and check whether str_end points to the terminating null character.

:+1:

> See also my comment in ros2/rclcpp#1188 (comment). rclcpp obviously has its own implementation of rcl_get_default_domain_id() and does not use this one.

Originally `rcl_get_default_domain_id` function didn't exist, thus it wasn't used in some places.

> See also my comment in ros2/rclcpp#1188 (comment). rclcpp obviously has its own implementation of rcl_get_default_domain_id() and does not use this one.

Actually, `domain_id` in `NodeOptions` is completely ignored now.
I have to delete that field in `rcl` and `rmw` layers (see https://github.com/ros2/rclcpp/issues/910). Luckily it was never exposed to the user in the upper layers.